File: pipeline_example/analysis.R
# analysis.R
#
# This script performs the analysis tasks for Question 3 of the case study.
# It is designed to be run after the Python data extraction script.
#
# Tasks:
# 1. Visualize daily average wind power density maps for Europe, faceted by day.
# 2. Compute country-level average wind power density for the 3-day run.
# 3. Rank countries from highest to lowest wind power potential.

# --- 1. Setup and Package Installation ---
# List of required packages
required_packages <- c(
  "RSQLite",    # For reading from SQLite database
  "dplyr",      # For data manipulation
  "lubridate",  # For date-time manipulation
  "ggplot2",    # For plotting
  "sf",         # For handling spatial vector data (countries)
  "terra",      # For handling spatial raster data
  "rnaturalearth", # For country boundary data
  "exactextractr", # For summarizing raster data over polygons
  "viridis"     # For color palettes
)

# Function to install packages if they are not already installed
install_if_missing <- function(pkg) {
  if (!require(pkg, character.only = TRUE)) {
    install.packages(pkg, repos = "http://cran.us.r-project.org")
  }
  library(pkg, character.only = TRUE)
}

# Apply the function to all required packages
suppressPackageStartupMessages(sapply(required_packages, install_if_missing))

# --- 2. Data Loading and Preparation ---
args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 2) {
  stop("Usage: Rscript analysis.R <YYYYMMDD> <CYCLE>", call. = FALSE)
}
date_str <- args[1]
cycle_str <- args[2]

db_path <- "data/processed/gfs_data.db"
if (!file.exists(db_path)) {
  stop("Database file not found. Please run the Python data extractor first.", call. = FALSE)
}

# Connect to the SQLite database and load data
con <- RSQLite::dbConnect(RSQLite::SQLite(), db_path)
query <- sprintf(
  "SELECT lat, lon, wind_power_density, forecast_hour FROM gfs_forecasts WHERE forecast_date = '%s' AND cycle = '%s'",
  date_str, cycle_str
)
gfs_data <- RSQLite::dbGetQuery(con, query)
RSQLite::dbDisconnect(con)

if (nrow(gfs_data) == 0) {
  stop("No data found in the database for the specified date and cycle.", call. = FALSE)
}

# --- 3. Task 2b: Visualize Daily Average Wind Power Density (Faceted) ---

# Convert forecast hours into specific dates
gfs_data_daily <- gfs_data %>%
  mutate(
    run_date = as.Date(date_str, format = "%Y%m%d"),
    forecast_datetime = run_date + hours(forecast_hour),
    forecast_day = as.Date(forecast_datetime)
  )

# Calculate the average wind power density for each day and grid cell
daily_avg_wpd <- gfs_data_daily %>%
  group_by(lat, lon, forecast_day) %>%
  summarise(avg_wpd = mean(wind_power_density, na.rm = TRUE), .groups = 'drop')




# Get European country boundaries for the map overlay
europe_countries <- ne_countries(scale = "medium", returnclass = "sf", continent = "Europe")

# Create the faceted plot
wpd_map_faceted <- ggplot() +
  geom_raster(data = daily_avg_wpd, aes(x = lon, y = lat, fill = avg_wpd)) +
  geom_sf(data = europe_countries, fill = NA, color = "black", linewidth = 0.2) +
  # Use a better projection for Europe (Lambert Azimuthal Equal-Area)
  coord_sf(crs = "+proj=laea +lat_0=52 +lon_0=10 +x_0=4321000 +y_0=3210000 +ellps=GRS80 +units=m",
           xlim = c(2.5e6, 6.5e6), ylim = c(1.5e6, 5.5e6)) +
  # Use a more appropriate color scale for wind
  scale_fill_viridis_c(
    name = "Wind Power Density (W/m²)",
    option = "plasma",
    limits = c(0, max(daily_avg_wpd$avg_wpd, na.rm = TRUE)),
    na.value = "transparent"
  ) +
  # Create a separate map for each day
  facet_wrap(~ forecast_day, ncol = 3) +
  labs(
    title = paste("Daily Average Wind Power Density Forecast"),
    subtitle = paste("GFS Run:", date_str, "Cycle", cycle_str),
    x = NULL, y = NULL
  ) +
  theme_minimal() +
  theme(
    legend.position = "bottom",
    legend.key.width = unit(2, "cm"),
    plot.title = element_text(size = 16, face = "bold"),
    strip.text = element_text(size = 12, face = "bold")
  )

# Save the improved map
output_dir <- "data/processed/plots"
dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)
ggsave(
  filename = file.path(output_dir, paste0("wpd_map_faceted_", date_str, "_", cycle_str, ".png")),
  plot = wpd_map_faceted,
  width = 15,
  height = 7
)

cat("Successfully generated and saved faceted wind power density map.\n")

# --- 4. Task 3: Country Ranking by Wind Power Density ---

# First, calculate the average WPD over the entire 3-day run for each grid cell
total_avg_wpd <- gfs_data %>%
  group_by(lat, lon) %>%
  summarise(avg_wpd = mean(wind_power_density, na.rm = TRUE), .groups = 'drop')

# Convert to a raster for spatial analysis
total_wpd_raster <- terra::rast(total_avg_wpd, type = "xyz", crs = "EPSG:4326")

# Use exact_extract to get the mean WPD for each country over the full forecast period
country_avg <- exact_extract(total_wpd_raster, europe_countries, "mean")

# Combine the results with the country names and rank them
country_results <- europe_countries %>%
  select(name, iso_a3) %>%
  mutate(avg_wpd_3day = country_avg) %>%
  st_drop_geometry() %>%
  filter(!is.na(avg_wpd_3day)) %>%
  arrange(desc(avg_wpd_3day)) %>%
  mutate(rank = row_number())

# Display the results
cat("\n--- Country Wind Power Density Rankings (3-Day Average) ---\\n")
print(head(country_results, 20))

# Save the rankings to a CSV file
output_csv_path <- file.path(output_dir, paste0("country_rankings_", date_str, "_", cycle_str, ".csv"))
write.csv(country_results, output_csv_path, row.names = FALSE)
cat(paste("\nSuccessfully saved country rankings to", output_csv_path, "\n"))

# Save the rankings to the database
con <- RSQLite::dbConnect(RSQLite::SQLite(), db_path)

# --- Delete existing rankings for this forecast cycle to avoid duplicates ---
delete_query <- sprintf(
  "DELETE FROM country_rankings WHERE forecast_date = '%s' AND cycle = '%s'",
  date_str, cycle_str
)
RSQLite::dbExecute(con, delete_query)
cat(sprintf("\nDeleted existing rankings for %s cycle %s.\n", date_str, cycle_str))

db_rankings <- country_results %>%
  mutate(forecast_date = date_str, cycle = cycle_str) %>%
  rename(country = name, avg_wind_power_density = avg_wpd_3day) %>%
  select(forecast_date, cycle, country, avg_wind_power_density, rank)

# Ensure table exists and then append
if (!"country_rankings" %in% RSQLite::dbListTables(con)) {
    RSQLite::dbCreateTable(con, "country_rankings", db_rankings)
}
RSQLite::dbWriteTable(con, "country_rankings", db_rankings, append = TRUE, row.names = FALSE)
RSQLite::dbDisconnect(con)


cat("Successfully saved country rankings to the database.\n")

File: pipeline_example/src/analysis.py
import duckdb
import pandas as pd
import geopandas as gpd
from shapely.geometry import Point
import matplotlib.pyplot as plt
import os
import requests
import zipfile
from pathlib import Path

def get_country_shapes():
    """
    Downloads and returns country shapes for Europe.
    """
    # Define the path for storing the shapefile
    data_dir = Path("data/shapefiles")
    data_dir.mkdir(parents=True, exist_ok=True)
    
    shapefile_path = data_dir / "ne_110m_admin_0_countries.shp"
    
    # Download the shapefile components if they don't exist
    if not shapefile_path.exists():
        print("Downloading Natural Earth countries shapefile components...")
        
        # List of required shapefile components
        base_url = "https://github.com/nvkelso/natural-earth-vector/raw/master/110m_cultural/"
        files_to_download = [
            "ne_110m_admin_0_countries.shp",
            "ne_110m_admin_0_countries.shx", 
            "ne_110m_admin_0_countries.dbf",
            "ne_110m_admin_0_countries.prj",
            "ne_110m_admin_0_countries.cpg"
        ]
        
        for filename in files_to_download:
            file_path = data_dir / filename
            if not file_path.exists():
                print(f"Downloading {filename}...")
                url = base_url + filename
                response = requests.get(url)
                response.raise_for_status()
                
                with open(file_path, 'wb') as f:
                    f.write(response.content)
        
        print("Shapefile components downloaded successfully.")
    
    # Read the shapefile
    world = gpd.read_file(shapefile_path)
    europe = world[world['CONTINENT'] == 'Europe']
    return europe

def create_europe_dashboard_data(db_path, date_str, cycle_str, output_dir):
    """
    Generates data for the Europe dashboard.
    """
    con = duckdb.connect(db_path)
    query = f"""
    SELECT lat, lon, wind_power_density, forecast_hour 
    FROM gfs_forecasts 
    WHERE forecast_date = '{date_str}' AND cycle = '{cycle_str}'
    """
    gfs_data = con.execute(query).fetchdf()
    con.close()

    if gfs_data.empty:
        print("No data found for the specified date and cycle.")
        return

    gfs_data['forecast_datetime'] = pd.to_datetime(date_str) + pd.to_timedelta(gfs_data['forecast_hour'], unit='h')
    gfs_data['forecast_day'] = gfs_data['forecast_datetime'].dt.date

    daily_avg_wpd = gfs_data.groupby(['lat', 'lon', 'forecast_day'])['wind_power_density'].mean().reset_index()

    europe = get_country_shapes()

    fig, ax = plt.subplots(1, 1, figsize=(20, 10))
    europe.boundary.plot(ax=ax)
    
    for day in daily_avg_wpd['forecast_day'].unique():
        day_data = daily_avg_wpd[daily_avg_wpd['forecast_day'] == day]
        plt.scatter(day_data['lon'], day_data['lat'], c=day_data['wind_power_density'], cmap='viridis', s=1)

    plt.title(f"Daily Average Wind Power Density (GFS Run: {date_str} Cycle {cycle_str})")
    plt.xlabel("Longitude")
    plt.ylabel("Latitude")
    
    plot_path = os.path.join(output_dir, f"wpd_map_faceted_{date_str}_{cycle_str}.png")
    plt.savefig(plot_path)
    plt.close()  # Close the figure to free memory
    print(f"Successfully generated and saved faceted wind power density map to {plot_path}")

    total_avg_wpd = gfs_data.groupby(['lat', 'lon'])['wind_power_density'].mean().reset_index()
    geometry = [Point(xy) for xy in zip(total_avg_wpd['lon'], total_avg_wpd['lat'])]
    points_gdf = gpd.GeoDataFrame(total_avg_wpd, geometry=geometry, crs="EPSG:4326")

    joined_gdf = gpd.sjoin(points_gdf, europe, how="inner", predicate='within')
    # Use the correct column name for country names in Natural Earth data
    country_column = 'NAME' if 'NAME' in joined_gdf.columns else 'name'
    country_avg = joined_gdf.groupby(country_column)['wind_power_density'].mean().reset_index()
    country_avg = country_avg.sort_values(by='wind_power_density', ascending=False).reset_index(drop=True)
    country_avg['rank'] = country_avg.index + 1
    
    csv_path = os.path.join(output_dir, f"country_rankings_{date_str}_{cycle_str}.csv")
    country_avg.to_csv(csv_path, index=False)
    print(f"Successfully saved country rankings to {csv_path}")


def create_specific_points_dashboard_data(db_path, date_str, cycle_str, coordinates_file, output_dir):
    """
    Generates data for the specific points dashboard.
    """
    con = duckdb.connect(db_path)
    query = f"""
    SELECT lat, lon, wind_power_density, forecast_hour 
    FROM gfs_forecasts 
    WHERE forecast_date = '{date_str}' AND cycle = '{cycle_str}'
    """
    gfs_data = con.execute(query).fetchdf()
    con.close()

    if gfs_data.empty:
        print("No data found for the specified date and cycle.")
        return

    coordinates = pd.read_csv(coordinates_file)

    gfs_data['geometry'] = [Point(xy) for xy in zip(gfs_data['lon'], gfs_data['lat'])]
    gfs_gdf = gpd.GeoDataFrame(gfs_data, geometry=gfs_data['geometry'], crs="EPSG:4326")

    coordinates['geometry'] = [Point(xy) for xy in zip(coordinates['lon'], coordinates['lat'])]
    coord_gdf = gpd.GeoDataFrame(coordinates, geometry=coordinates['geometry'], crs="EPSG:4326")

    joined_gdf = gpd.sjoin_nearest(coord_gdf, gfs_gdf, how="inner")
    
    specific_points_avg = joined_gdf.groupby('site_name')['wind_power_density'].mean().reset_index()
    specific_points_avg = specific_points_avg.sort_values(by='wind_power_density', ascending=False).reset_index(drop=True)
    specific_points_avg['rank'] = specific_points_avg.index + 1

    csv_path = os.path.join(output_dir, f"specific_points_rankings_{date_str}_{cycle_str}.csv")
    specific_points_avg.to_csv(csv_path, index=False)
    print(f"Successfully saved specific points rankings to {csv_path}")


def main(date_str, cycle_str):
    db_path = "data/processed/gfs_data.duckdb"
    output_dir = "data/processed/plots"
    coordinates_file = "wind_power_plants_coordinates.csv"
    
    os.makedirs(output_dir, exist_ok=True)

    create_europe_dashboard_data(db_path, date_str, cycle_str, output_dir)
    create_specific_points_dashboard_data(db_path, date_str, cycle_str, coordinates_file, output_dir)

if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument("--date", required=True)
    parser.add_argument("--cycle", required=True)
    args = parser.parse_args()
    main(args.date, args.cycle)
